container/image_test.py

```diff
@@ -336,6 +333,96 @@ def test_py_image(self):
         './app/io_bazel_rules_docker/testdata',
         './app/io_bazel_rules_docker/testdata/py_image_library.py',
         './app/io_bazel_rules_docker/testdata/__init__.py',
+        '/app',
+        '/app/testdata',
+        '/app/testdata/py_image.binary.runfiles',
+        '/app/testdata/py_image.binary.runfiles/io_bazel_rules_docker',
+        '/app/testdata/py_image.binary.runfiles/io_bazel_rules_docker/testdata',
+        '/app/testdata/py_image.binary.runfiles/io_bazel_rules_docker/testdata/py_image_library.py',
+        '/app/testdata/py_image.binary.runfiles/io_bazel_rules_docker/testdata/__init__.py',
+      ])
```

Comment:
Putting the symlinks for all layers into the binary's layer was (subtle but) intentional because it enables the same dependency's layer to be shared by multiple binary targets.

Looking at the "before" diff you see no mention of `py_image.binary` in the files, this means that I could have two `py_image` targets `foo` and `bar` consume the same dependency and both would utilize the same layer for that dependency in the registry (eliding uploads and downloads).

The concern isn't space in the registry (e.g. avoiding N copies of a `grpcio` dependency) it's more about cache density.  The Registry is effectively a distributed cache and if we make the layer for the same dependency always produce the same layer, then all consumers of that layer hit once one of them publishes it.  The same is true of pull latency, since the daemon shouldn't repull a layer it already has.

Reply:
It is notable that if we're going to head this route, we could also simply forgo the symlinks altogether, right?

Reply:
> It is notable that if we're going to head this route, we could also simply forgo the symlinks altogether, right?

Yes. I wonder if we could just expose `agnostic_dep_layout` attribute in `py_image`...? Then the user could pick from the following combinations:

`layering_strategy='group_external' agnostic_dep_layout=False`
- Create external layer, no symlinks, no cache sharing between images

`layering_strategy='group_external' agnostic_dep_layout=True`
- Create "external layer", and create "external symlinks layer". The latter would create symlinks to the former for the runfiles dir of the current image. The external layer could be shared across images, the symlinks layer not.

`layering_strategy='default' agnostic_dep_layout=False`
- Current behavior of `py_image`, except don't create symlinks, no cache sharing between images. This can still reduce symlink overhead in the app layer, even if you don't want to lump all external deps to a single layer.

`layering_strategy='default' agnostic_dep_layout=True`
- Current behavior of `py_image`.

Reply:
I can't speak for Matt, but I believe that I have addressed his concern in the latest version of the PR. In my initial version (it's a shame that force-push has lost that..), I had combined external files and external symlinks in a single "external layer", and Matt correctly pointed out that such layers could not be reused between different binary targets. However, in the latest version, we no longer have such "external layer" that would mix both files and symlinks. Instead, we have separate layers for *files* that can be reused across binary targets (layers with `./app/*`), and separate layers for *syminks* that cannot be shared (layers with `/app/testdata/py_image.binary.runfiles/*`).

The difference from the current top-of-tree is that my PR does not lump *all* symlinks to the top layer. This is to split the top layer (15MB in my tensorflow example) to stable symlinks (e.g. symlinks to tensorflow-gpu files, change rarely) and app files (change frequently) and hence reduce build+push delta.

Reply:
sgtm